File: external/Java.Interop/src/Java.Interop/java/com/xamarin/android/ManagedPeer.java
package com.xamarin.android;

public /* static */ class ManagedPeer {
	private ManagedPeer () {
	}

	// public static native void registerNativeMethods (java.lang.Class<?> nativeClass, String managedType, String methods);

	public static void runConstructor (
			Class<?> declaringClass,
			Object self,
			String assemblyQualifiedName,
			String constructorSignature,
			Object... arguments) {
		if (self.getClass() != declaringClass)
			return;
		runConstructor (self, assemblyQualifiedName, constructorSignature, arguments);
	}

	static native void runConstructor (
			Object self,
			String assemblyQualifiedName,
			String constructorSignature,
			Object... arguments
	);
}

File: external/Java.Interop/src/Java.Interop/Java.Interop/ManagedPeer.cs
using System;
using System.Diagnostics;
using System.Reflection;
using System.Text;

namespace Java.Interop {

	[JniTypeInfo (ManagedPeer.JniTypeName)]
	/* static */ class ManagedPeer : JavaObject {

		internal const string JniTypeName = "com/xamarin/android/ManagedPeer";


		static  readonly    JniPeerMembers  _members        = new JniPeerMembers (JniTypeName, typeof (ManagedPeer));

		static ManagedPeer ()
		{
			_members.JniPeerType.RegisterNativeMethods (
					new JniNativeMethodRegistration ("runConstructor",  RunConstructorSignature,    (Action<IntPtr, IntPtr, IntPtr, IntPtr, IntPtr, IntPtr>) RunConstructor)
			);
		}

		ManagedPeer ()
		{
		}

		internal static void Init ()
		{
			// Present so that JavaVM has _something_ to reference to
			// prompt invocation of the static constructor & registration
		}

		public override JniPeerMembers JniPeerMembers {
			get {return _members;}
		}

		const string RunConstructorSignature = "(Ljava/lang/Object;Ljava/lang/String;Ljava/lang/String;[Ljava/lang/Object;)V";

		static void RunConstructor (
				IntPtr jnienv,
				IntPtr klass,
				IntPtr n_self,
				IntPtr n_assemblyQualifiedName,
				IntPtr n_constructorSignature,
				IntPtr n_constructorArguments)
		{
		}
	}
}

